### PR TITLE
Refactor and add Decorator for API 4xx Exception Views

### PR DESCRIPTION
**Pull request by lyzadanger**
This PR makes no user-facing or functional changes, but restructures the API's exception views in a manner that will make it possible to sub in 415s instead of 404s in some cases. I almost just added that last part into this PR but I didn't want the diff to get too big.

What's in here:

* Instead of messing with the HTTP exception directly in the exception view, there is now a decorator that will alter the `context` passed to the view. When I'm done with this whole process (one more PR, I'd expect), the exception view will expect some form of `HTTPClientError` as its `context` but doesn't need to know more than that.
* Turning off `debug_all` for functional tests because it accomplishes nothing and interferes with the tests I started adding here...
* Functional tests for error response details so I can be sure I'm not inadvertently changing current behavior.
* Stop stringifying the whole `context` in the more generic `HTTPError` handler (which will be conflated into a single exception view in a later PR instead of being standalone).

Basically this is a cleanup and prep PR so I can add the functionality I've actually been trying to add for 2 or 3 days now :). That next part I have a pretty good handle on!

### CODE REVIEW

**Comment by lyzadanger** [comment]
Looks like I've got some string-encoding issues in my functional tests (they run fine locally, of course)...I'll dive into that after review.

File: h/views/api/decorators/client_errors.py [hunk 1]
```diff
@@ -0,0 +1,50 @@
+# -*- coding: utf-8 -*-
+from __future__ import unicode_literals
+
+from pyramid.httpexceptions import HTTPForbidden, HTTPNotFound
+
+
+def client_error_decorator(wrapped):
+    """
+    Decorate client (HTTP 4xx) exception views
```
Comment (seanh) [inline h/views/api/decorators/client_errors.py]: I was wondering about the name of this decorator and the first line of its docstring. "Decorate client error views" seems quite generic and describes where the decorator is used not what it does.

I wonder if it might be better to name the decorator after what it does. And that might also suggest splitting this into two separate decorators that each do one thing. Something like `@forbidden_view_config(..., decorator=(cors_policy, change_fordidden_to_notfound, change_notfound_to_unsupported_media_type)):`

File: h/views/api/decorators/client_errors.py [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+
+        # If we're dealing with a 403, substitute with a 404 instead
+        # FIXME: We should be more nuanced about when we do this
+        if isinstance(context, HTTPForbidden):
+            context = HTTPNotFound()
+
+        if isinstance(context, HTTPNotFound):
+            # The default message is not helpful: it's the path
+            # that failed to match any view. Replace it.
+            context.message = "Either the resource you requested doesn't exist, or you are not currently authorized to see it."
```
Comment (seanh) [inline h/views/api/decorators/client_errors.py]: I wonder if this belongs in the `@notfound_view_config` itself? The `if isinstance(context, HTTPNotFound):` especially makes me thing it belongs in there

File: tests/h/views/api/decorators/client_errors_test.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+
+    def test_it_does_not_replace_other_http_4xx_messages(
+        self, pyramid_request, testview
+    ):
+        wrapped = client_error_decorator(testview)
+        default_gone = HTTPGone()
+
+        wrapped(HTTPGone(), pyramid_request)
+
+        (context, _) = testview.call_args[0]
```
Comment (seanh) [inline tests/h/views/api/decorators/client_errors_test.py]: I don't think the `(...)`'s are needed here (and you wouldn't normally put them in)

File: tests/functional/api/test_errors.py [hunk 4]
```diff
@@ -0,0 +1,195 @@
...
+    }
+
+
+@pytest.fixture
+def user_with_token(db_session, factories):
+    user = factories.User()
+    token = factories.DeveloperToken(userid=user.userid)
+    db_session.add(token)
+    db_session.commit()
+    return (user, token)
```
Comment (seanh) [inline tests/functional/api/test_errors.py]: Normally if I had two different objects to create (a user and a token) I'd make two separate fixtures for them:

```python
@pytest.fixture
def user(db_session, factories):
    user = factories.User()
    db_session.commit()
    return user

@pytest.fixture
def token(db_session, factories, user):
    token = factories.DeveloperToken(userid=user.userid)
    db_session.commit()
    return token
```

The advantage is that any test or other fixture that wants to use one or the other of `user` and `token` can just take the one it wants as an argument, and not have to take both. So `token_auth_header()` below wouldn't have to do `user, token = user_with_token` and wouldn't have an unused variable `user`.

On the other hand it does mean `db_session.commit()` gets called twice but I think there's a general suspect-ness about test fixtures having to explicitly call `commit()` anyway

File: tests/functional/api/test_errors.py [hunk 5]
```diff
@@ -0,0 +1,195 @@
...
+            "Basic {creds}".format(creds=encoded.decode("ascii"))
+        )
+    }
+
+
+@pytest.fixture
+def user_with_token(db_session, factories):
+    user = factories.User()
+    token = factories.DeveloperToken(userid=user.userid)
+    db_session.add(token)
```
Comment (seanh) [inline tests/functional/api/test_errors.py]: I don't think `add()` is necessary here, creating a model object with a factory automatically adds it to the DB session

File: tests/functional/api/test_errors.py [hunk 6]
```diff
@@ -0,0 +1,195 @@
...
+        res = app.post_json(
+            "/api/groups", group, headers=auth_client_header, expect_errors=True
+        )
+
+        assert res.status_code == 404
+
+    def test_it_returns_formatted_reason_if_no_authenticated_user(
+        self, app, auth_client_header
+    ):
+        # FIXME: This should return a 403
```
Comment (seanh) [inline tests/functional/api/test_errors.py]: Is this FIXME covered by the xfail test below?

File: tests/functional/api/test_errors.py [hunk 7]
```diff
@@ -0,0 +1,195 @@
...
+        )
+
+
+@pytest.mark.functional
+class Test403ErrorsUsingCreateGroup(object):
+    # FIXME: API currently converts all 403s to 404s
+    def test_it_returns_http_404_if_no_authenticated_user(
+        self, app, auth_client_header
+    ):
+        # FIXME: This should return a 403
```
Comment (seanh) [inline tests/functional/api/test_errors.py]: Is this FIXME covered by the xfail test below?

File: tests/functional/api/test_errors.py [hunk 8]
```diff
@@ -0,0 +1,195 @@
...
+
+        assert (
+            res.json["reason"]
+            == "groupid: u'3434kjkjk' does not match u\"^group:([a-zA-Z0-9._\\\\-+!~*()']{1,1024})@(.*)$\""
+        )
+
+
+@pytest.mark.functional
+class Test403ErrorsUsingCreateGroup(object):
+    # FIXME: API currently converts all 403s to 404s
```
Comment (seanh) [inline tests/functional/api/test_errors.py]: I wonder if this FIXME belongs in a comment or docstring on the xfail test below?

File: tests/functional/api/test_errors.py [hunk 9]
```diff
@@ -0,0 +1,195 @@
...
+        res = app.post_json(
+            "/api/groups", group, headers=token_auth_header, expect_errors=True
+        )
+
+        assert res.status_code == 400
+
+    def test_it_returns_formatted_reason_if_groupid_set_on_default_authority(
+        self, app, token_auth_header
+    ):
+        # FIXME: The `reason` is double-escaped
```
Comment (seanh) [inline tests/functional/api/test_errors.py]: I'd move this close to the double-escaped reason string below

File: h/views/api/decorators/client_errors.py (hunk 1 above)
Comment (lyzadanger) [reply h/views/api/decorators/client_errors.py]: This is a really good idea. I'm going to split them up.

**Comment by lyzadanger** [comment]
Based on a suggestion in review, I've refactored and extended this branch just enough that I think it warrants opening a new PR instead of overloading this one.
